plugins/catalog-backend/src/permissions/rules/hasAnnotation.ts

```diff
@@ -23,7 +23,7 @@ import { CatalogPermissionRule } from '../types';
  * annotation on a given entity.
  * @public
  */
-export const hasAnnotation: CatalogPermissionRule = {
+export const hasAnnotation: CatalogPermissionRule<[annotation: string]> = {
```

Comment:
What do you think about providing a `createPermissionRule` so that we can infer these rather than having to duplicate? Possibly even a `makePermissionRuleFactory` that encapsulates the `Entity` and `EntitiesSearchFilter` params and returns a `createPermissionRule`

Reply:
That's actually how these rules were defined way back in the early days of the permissions PRFC, and I still like it! We ended up with a type param because it was uncomfortable having a function with no runtime responsibilities - it ends up looking like this:

```typescript
export const createPermissionRule = <
  TResource,
  TQuery,
  TParams extends unknown[],
>(
  rule: PermissionRule<TResource, TQuery, TParams>,
) => rule;
```

```typescript
export const makeCreatePermissionRule =
  <TResource, TQuery>() =>
  <TParams extends unknown[]>(
    rule: PermissionRule<TResource, TQuery, TParams>,
  ) =>
    createPermissionRule(rule);
```

```typescript
export const createCatalogPermissionRule = makeCreatePermissionRule<
  Entity,
  EntitiesSearchFilter
>();
```

Open to reverting to this approach! If it keeps the public API cleaner, it's perhaps preferable even if the internals are a bit strange.

Reply:
Yeah in TS there's no shame in having [functions](https://github.com/backstage/backstage/blob/bf7a73153111ee4b15dd8473f6f5a0793ccb0cfb/plugins/scaffolder-backend/src/scaffolder/actions/createTemplateAction.ts#L19) [that](https://github.com/backstage/backstage/blob/bf7a73153111ee4b15dd8473f6f5a0793ccb0cfb/packages/cli/src/commands/install/types.ts#L74) [don't](https://github.com/backstage/backstage/blob/bf7a73153111ee4b15dd8473f6f5a0793ccb0cfb/packages/cli/src/lib/create/types.ts#L73) [do anything](https://github.com/backstage/backstage/blob/bf7a73153111ee4b15dd8473f6f5a0793ccb0cfb/packages/core-plugin-api/src/apis/system/helpers.ts#L58) x)

Yeah the last one does a tiny but put largest chunk is still type inference :p
The second last is probably the most extreme, as it doesn't even return an inferred type, it's just to help out with the declaration.

Reply:
Sold! I'll push this change.

Reply:
Alright - a few changes here, keen to get your thoughts:
* b20c45384d adds helpers in permission-node, I think fairly uncontroversial
* a2b6418cf5 uses the helpers in catalog-backend. I think _mostly_ uncontroversial, but one thing that sticks out to me is that strictly speaking the rules are no longer of type `CatalogPermissionRule<TParams>` - they're now of type `PermissionRule` (with the appropriate params). This is a really minor issue, but it does bother me a bit - especially since we're referring to them as `CatalogPermissionRule`s in the doc comments.
* To address the issues in the previous commit, I've added a custom `createPermissionRule` implementation in catalog-backend, which does the same job as the on in permission-node, but uses the `CatalogPermissionRule` type. This means that in this particular case we don't actually use the helpers from permission-node any more.

An alternative to this last thing might be to discard the `CatalogPermissionRule` type entirely, and instead use the appropriately-parameterised `PermissionRule` type in the handful of places we explicitly refer to it.

Reply:
👍

From what I can tell the need for `CatalogPermissionRule` is not unique, so it'd be a bit of a shame if you always end up having to created your own rule creation utility, kinda defeats the purpose of the common one. I'm seeing 3 different options:

- Keep the utility local, i.e. declare `createCatalogPermissionRule`-like helpers wherever needed, and drop the common `makeCreatePermissionRule` utility
- Drop `CatalogPermissionRule`, use the more verbose `PermissionRule<Entity, EntitiesSearchFilter, ?>` wherever needed, it's just a shorthand after all
- Modify `makeCreatePermissionRule` so that it accepts a `CatalogPermissionRule`-like type rather than the individual parameters. The `CatalogPermissionRule` would no longer hold the parameters but just be a container for what resource and query types are expected, which I think makes sense with the usage in the builder. `makeCreatePermissionRule` could look something like this:

  ```ts
  export function makeCreatePermissionRule<TRule extends PermissionRule<unknown, unknown, unknown[]>>(): 
    TRule extends PermissionRule<infer TResource, infer TQuery>
      ? <TParams extends unknown[]>(rule: PermissionRule<TResource, TQuery, TParams>) => 
        PermissionRule<TResource, TQuery, TParams>
      : never {
    return ((rule: any) => rule) as any;
  }
```

Reply:
> From what I can tell the need for CatalogPermissionRule is not unique, so it'd be a bit of a shame if you always end up having to created your own rule creation utility, kinda defeats the purpose of the common one.

Agree - I think it's a matter of preference, and my intent here was to offer `makeCreatePermissionRule` for folks who don't want to have a custom alias for their parameterised `PermissionRule` type, while still finding a solution that works for folks who do.

> Modify makeCreatePermissionRule so that it accepts a CatalogPermissionRule-like type rather than the individual parameters. The CatalogPermissionRule would no longer hold the parameters but just be a container for what resource and query types are expected, which I think makes sense with the usage in the builder. makeCreatePermissionRule could look something like this `[snip]`

I did explore accepting types like CatalogPermissionRule rather than TResource and TQuery in `makeCreatePermissionRule`, and arrived at a similar implementation to yours. Unfortunately, I think this approach has the same issue as the existing one - it returns the type `PermissionRule<Entity, EntitiesSearchFilter, [exampleParam: string]>`, rather than `CatalogPermissionRule<[exampleParam: string]>`, which is the same problem we already have with the existing implementation of this function.

I'm not sure this is something we even need to sweat over - perhaps it's fine that we have a type `CatalogPermissionRule<TParams>` that's equivalent to `PermissionRule<Entity, EntitiesSearchFilter, TParams>` that's available for convenience but isn't used everywhere. We can get to that state by reverting https://github.com/backstage/backstage/commit/0341e562341a7296baa4ea9f01e2a811d1e6ceb1.

Reply:
Had a bit of a chat offline with @Rugvip and agreed that we'd revert 0341e56 and drop the CatalogPermissionRule type in favor of providing a parameterised `PermissionRule` type in the few places we need it inside catalog-backend. This is nice, because it means we don't have this additional public type exported by the package which is probably not appropriate for external use, but would be tempting to use.

Reply:
Done in 964e58ddec.